Argparse integration (continued)

stephenfin opened the pull request:
Subject: Continued argparse-ification of the Sphinx utilities.

### Feature or Bugfix
Refactor

### Purpose
This continues the work on argparse, cleaning up some stuff that wasn't included in #4097 so as to keep it as simple as possible. Refer to #3259 and #4097 for more information.

### Detail
- quickstart: Rework how we collect extensions

  This is a little more flexible than the existing set up and makes maximum use of argparse capabilities.

  This has the side-effect of no longer including configuration for the `sphinx.ext.todo` extension when said extension is not enabled.

This requires two additional changes.

- quickstart: Rework 'do_prompt' function

  The 'd' and 'key' values are used on a single line. Move these outside the function to allow us to do other things with this function.

- quickstart: Simplify `--(no-)option` handling

  The standard pattern for doing this is to use a common 'dest'.

### Relates
- #3259 
- #4097

stephenfin committed:
quickstart: Simplify --(no-)option handling

The standard pattern for doing this is to use a common 'dest'.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
quickstart: Rework 'do_prompt' function

The 'd' and 'key' values are used on a single line. Move these outside
the function to allow us to do other things with this function.

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented on sphinx/cmd/quickstart.py:
```diff
@@ -48,24 +49,43 @@
+    ('texinfo',
+     'generate Texinfo files that can be processed into Info files'),
+    ('epub', 'generate an EPUB file for e-book readers'),
+])
```
How did you choice these builders?

stephenfin replied:
These are the builders for which there was configuration in the existing quickstart template. I can add more but that should be another patch/pull request, I think?

tk0miya commented on sphinx/cmd/quickstart.py:
```diff
@@ -260,7 +280,7 @@ def ask_user(d):
         print('sphinx-quickstart will not overwrite existing Sphinx projects.')
         print()
-        do_prompt(d, 'path', 'Please enter a new root path (or just Enter '
+        d['path'] = do_prompt('Please enter a new root path (or just Enter '
```
Good improvement! Could you fix indentation of hanging lines?

stephenfin replied:
Will do

tk0miya commented on sphinx/cmd/quickstart.py:
```diff
@@ -332,65 +352,42 @@ def ask_user(d):
+        d['builders'] = []
+        for name, description in BUILDERS:
+            if do_prompt('%s: %s (y/n)' % (name, description), 'n', boolean):
+                d['builders'].append(name)
```
Did you mean all builders in BUILDERS are disabled by default? Really?

stephenfin replied:
Not quite. Per one of the commit messages and the pull request summary:

> For some reason, we include the epub builder config from 'conf.py' in the list of "things users can modify", yet we don't do this for any other builder. Most users only care about the HTML builder and everything else is noise. As such, we should make this configurable and default all non-HTML builders to False, with the option to override this behavior.

As a result, the `html` builder is always enabled and other builders are disabled and most people don't use them. This is the same thing we do for extensions (enable them only if you actually want them)

tk0miya replied:
HTML builder is only a part of Sphinx. Other builders are also important for Sphinx. So I opposite for your proposal.

stephenfin replied:
Right, but I'm not removing anything: the users will still get asked if they want to use a builder or not (see below). As it is, the `conf.py` file already includes a lot of configuration options we really don't need. However, even with the current size of the file, we still don't include configuration for _extensions_ we aren't using. Why include configuration for _builders_ we aren't using? Surely if a user wanted to build PDFs or generate man pages then they would say so?

---

    htmlhelp: generate HTML help support files for Microsoft HTML Help Workshop (y/n)
    latex: generate LaTeX files that can be processed into PDF and other formats (y/n)
    man: generate man pages in groff format (y/n)
    texinfo: generate Texinfo files that can be processed into Info files (y/n)
    epub: generate an EPUB file for e-book readers (y/n)

tk0miya replied:
I agreed with current `conf.py` contains many options. And almost of them are commented out, so they are not important. But it does not means removing whole of them.
I believe some HTML options are also not needed.

I prefer to remove unneeded options and commented-out blocks from initial `conf.py`. It's acceptable for me.

stephenfin replied:
I'm not sure what you mean. It sounds like you agree with me, but think we should go further and remove a lot more by default? If this is the case, would this be an acceptable first step? I can focus on removing more of the noise, including some of the HTML builder options in another pull request? This pull request is focused on making builder configuration optional

tk0miya replied:
Sorry for confuse you. I said NO to make builder configuration optional.

stephenfin replied:
So what would you like me to do here? Drop the entirety of [the "configurable builder" change](https://github.com/sphinx-doc/sphinx/pull/4174/commits/2a709b1866557a1ae253d7a083cea7e20593afb4) or rework it to default every builder to enabled?

tk0miya replied:
I think dropping the commit is better.

stephenfin replied:
I've modified this to default to true. I understand that you consider other builders as very important so a default value of enabled makes sense. However, I also want to offer this option to users that don't actually need the other builders. Can you consider this acceptable?

tk0miya replied:
No. Personally, I proposed to reduce questions of quickstart. So your proposal conflicts with mine.

And we have to know the effect of the change. At least, we have to know how they works when all configurations are removed. And we should provide the way to enable them.

Are the options really important for users? I don't think so.

stephenfin replied:
So, let's make sure I'm getting this clear :smile: My problem is that this is all 90% of people need in their configuration file:

    master_doc = 'index'

That's it. If you're doing what most people do and using the `html` builder with no extensions to publish docs to readthedocs.org then nothing else is necessary. Compare this to the `conf.py` files that most projects have, like [this one](https://github.com/openstack/nova/blob/15.0.0/doc/source/conf.py). All those comments are noise and we don't need them. Instead we should simply be pointing people to docs if they want to configure stuff. This idea aims to tackle part of that by removing one thing that people almost certainly won't need: configuration for other builders.

I think you're suggesting that we ask ask *fewer questions* up front but have *more options* in the resulting `conf.py`? Meanwhile, I'm suggesting we ask *more questions* up front but have *fewer options** in the resulting `conf.py`. I think my option is a good one because you only run `sphinx-quickstart` once per project, whereas you might have to modify `conf.py` hundreds or thousands of times throughout a projects lifetime. In essence, the former approach is **lower initial cost** for **higher lifetime cost**. My proposal has **higher initial cost** for **lower lifetime cost**. Suggesting that users spend a little longer tweaking their `conf.py` is a reasonable ask. If they do not wish to do this then they can pass the ``--quiet`` option and ignore this whole process.

I really don't want to drop this because while you want a smaller number of options, I want a smaller `conf.py`. If you still don't agree with the above (and I urge you to think about it), what other options are available? Could I simply remove the `--epub` option and remove all non-HTML configuration options from `conf.py`? Maybe I could remove the HTML configuration options too? Maybe I could even go as far as removing *all* commented out options, and adding a link to the configuration guide in the comment at the top? Thoughts?

tk0miya replied:
I don't opposite to drop options from `conf.py`. I only opposite to make builders optional. We can reduce options without it.

Anyway, I feel very tired for reviewing your PR. So I'll stop maintainance work for a while.
Thanks,

stephenfin replied:
Oh, I'm sorry - I didn't mean to be rude and if I came across that way then I do apologize :disappointed: I'm trying to understand your objections so we can come to a compromise that makes everyone happy. This took a lot of my (personal) time and I don't want it to go to waste. I'm also trying my best to help review other patches and bugs and have a big documentation rework ready to go, to help you guys if possible. I'll drop this patch from the series for now, but I would like to address the whole "`conf.py` is too darn big" issue in the future. What would be the best way to discuss this?

tk0miya replied:
I'm back now. Sorry, I was very stupid. I must apologize you to my rudeness. I'd like to collaborate with you if you will forgive me.

To make a step forward, could you add a new issue to simplify conf.py please? I feel this PR contains a lot of elements; refactoring, interface for extensions and making builders optional. IMO, one issue one problem is better.

stephenfin replied:
It's no problem :+1: Maintaining stuff is stressful work for little reward, so totally understandable. I'm out of the country at the moment, but I'll take a look at this (and other comments) when I return in a week. Hopefully that's OK.

stephenfin replied:
@tk0miya This is noted in #4256. I have removed the offending patch here now. Is anything else needed? The only change here is to convert how we collect extensions that should be enabled.

tk0miya commented on sphinx/cmd/quickstart.py:
```diff
@@ -12,6 +12,7 @@
 from __future__ import absolute_import
 
 import argparse
+import collections
```
Personally, I prefer to import `OrderedDict` only.

stephenfin replied:
I'm following OpenStack coding standards which insist on [module, not object, imports](https://docs.openstack.org/hacking/latest/user/hacking.html#imports) as it results in less imports. Can fix though

stephenfin committed:
quickstart: Rework how we collect extensions

This is a little more flexible than the existing set up and makes
maximum use of argparse capabilities.

This has the side-effect of no longer including configuration for the
'sphinx.ext.todo' extension when said extension is not enabled.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin commented:
@tk0miya Is there anything else I need here?

tk0miya commented:
Merged.
Thank you for good and tough work!

stephenfin commented:
Thanks @tk0miya. Appreciated as always 😄
